### ISSUE TITLE
Provider Split and Modules

### ISSUE THREAD

**Issue by morgante** [terraform-providers/terraform-provider-google#2216]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

The new provider split has made it very hard to work with reusable modules because the GA `google` provider will error out if you include a beta feature in your config even if the module itself makes using that feature optional. This forces modules to only use the beta provider and beta APIs, even when they don't need to.

The beta provider should only be needed when you're actually changing those fields from their defaults.

### Potential Terraform Configuration

For example, a GKE module might want to have [Binary Authorization](https://www.terraform.io/docs/providers/google/r/container_cluster.html#enable_binary_authorization) be optional. The module code might look like this:

```tf
resource "google_container_cluster" "primary" {
  name               = "${var.cluster_name}
  zone               = "us-central1-a"
  initial_node_count = 3

  enable_binary_authorization = "${var.enable_binary_authorization}"
  }
}
```

This module will now require that all clusters be created with the beta API - even in cases where binary authorization is disabled.

### References

This is going to be an issue for virtually all of our open source modules, we just discussed it for GKE [here](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/13#issuecomment-428625323)

**Comment by paddycarver** [comment]
I think that's a fair point. On the one hand, prior to the provider split, we were forcing everything onto beta _anyways_, so that's not really a change here, it just means people have to download the beta provider. I'd argue that fields only sometimes being set is usually a smell of an overly-generic module, but that's probably a separate discussion.

I think we _may_ get a "free" fix for this with 0.12, with the use of `nil` for conditionally unset fields.

**Comment by morgante** [comment]
> > I'd argue that fields only sometimes being set is usually a smell of an overly-generic module, but that's probably a separate discussion.

We can discuss that separately, but I'd definitely disagree. We shouldn't force a whole new module just because somebody wants a slightly different config (ex. enabling binary authorization doesn't make your GKE cluster fundamentally different).

**Comment by paddycarver** [comment]
> Fair point, but if I'm in a module using the beta provider how do I make my resources reference that provider?

I think more testing is needed on this one, but I'd think a `provider = "${uses_beta_fields ? "google-beta" : "google"}"` would work?

> We can discuss that separately, but I'd definitely disagree.

I'd very much like to sit down and talk through this at some point. After all, if your module is just "a GKE cluster", that's... a resource? From what I understand, modules are intended to take the general-purpose resources and encode specific use cases. In that understanding, a module that conditionally may or may not use a beta feature feels too far on the generic side of the scale.

**Comment by morgante** [comment]
> From what I understand, modules are intended to take the general-purpose resources and encode specific use cases.

They're also meant to make the general resources easier to use and get started with, not just define a narrow use case. For example, all the top modules on the [registry](https://registry.terraform.io/) are general purpose utilities - not necessarily specific combinations. If that were the case, practically all the modules on the registry shouldn't exist.

We can catch up at Hashiconf, but I strongly feel we shouldn't force users to fork a module just to do something like activating a single beta field.

**Comment by paddycarver** [comment]
> They're also meant to make the general resources easier to use and get started with

That's a new take on the matter to me. Which definitely doesn't mean it's wrong, it just means my reasoning hasn't been including it. Thinking it through, I'm not sure pre-0.12 HCL (or even post-0.12 HCL) provides the tooling to do this effectively, but that's probably a longer talk at HashiConf.

> I strongly feel we shouldn't force users to fork a module just to do something like activating a single beta field.

I think what I'm imagining is less individual users forking, and more a different granularity in what opinions a module encodes.

I think we're largely just disagreeing over the level of abstraction that modules should provide.

The way I see it, in the short term, we have a few ways forward:

1. Get rid of the google-beta provider and reunite the providers into one. This is... a very long shot, and we should probably exhaust other options before we start discussing it, because that's a very hard sell.
2. Hardcode the `provider = "google-beta"` line into the resources that use beta, which may or may not work, and testing would have to be done to find out. But that means users can't use custom builds or specify which provider gets used or any of that fun stuff. Which may be fine.
3. Document that users need to pass the google-beta provider in using [explicit providers](https://www.terraform.io/docs/modules/usage.html#passing-providers-explicitly) when using modules using beta features.
4. A combination of 2 and 3, where you hardcode google-beta, and users can override that with explicit providers, which may work. I'm honestly not clear.
5. Things stay as they are now until 0.12 lands and we can get a permanent solution using null. I'm not sure what the situation currently looks like, so I don't know how problematic that is.
6. We add fields back to the GA provider to mirror the beta provider fields that throw an error if they're set to anything, but silently accept the empty string or whatever. This allows us to work around not having null support, but it's a really hard sell. I think you're more likely to see this than the providers merged again, but I think there'd need to be a pretty large fire going before I'd vote for this over just waiting for 0.12.

**Comment by rileykarson** [comment]
Is there any action we still need to take on this? You can see https://github.com/terraform-providers/terraform-provider-google/pull/2709 for an example of some of the interactions of the split with modules & honestly most modules should require using `google-beta` if they have any beta features.

**Comment by rileykarson** [comment]
There's nothing actionable here, so closing this out. If you have any concerns please raise them here / in a new issue @morgante.
